client/driver/lxc.go

```diff
@@ -0,0 +1,420 @@
+		if err := h.container.Stop(); err != nil {
+			h.logger.Printf("[ERR] driver.lxc: error stopping container %q: %v", h.container.Name(), err)
+		}
+	}
+	close(h.doneCh)
+	return nil
+}
+
+func (h *lxcDriverHandle) Stats() (*cstructs.TaskResourceUsage, error) {
+	cpuStats, err := h.container.CPUStats()
+	if err != nil {
+		return nil, nil
+	}
+	total, err := h.container.CPUTime()
+	if err != nil {
+		return nil, nil
+	}
+
+	t := time.Now()
+
+	// Get the cpu stats
+	system := cpuStats["system"]
+	user := cpuStats["user"]
+	cs := &cstructs.CpuStats{
+		SystemMode: h.systemCpuStats.Percent(float64(system)),
+		UserMode:   h.systemCpuStats.Percent(float64(user)),
+		Percent:    h.totalCpuStats.Percent(float64(total)),
+		TotalTicks: float64(user + system),
+		Measured:   LXCMeasuredCpuStats,
+	}
+
+	// Get the Memory Stats
+	memData := map[string]uint64{
+		"rss":   0,
+		"cache": 0,
+		"swap":  0,
+	}
+	rawMemStats := h.container.CgroupItem("memory.stat")
+	for _, rawMemStat := range rawMemStats {
+		key, val, err := keysToVal(rawMemStat)
+		if err != nil {
+			h.logger.Printf("[ERR] driver.lxc: error getting stat for key %q: %v", key, err)
+			continue
+		}
+		if _, ok := memData[key]; ok {
+			memData[key] = val
+
+		}
+	}
+	ms := &cstructs.MemoryStats{
+		RSS:      memData["rss"],
+		Cache:    memData["cache"],
+		Swap:     memData["swap"],
+		Measured: LXCMeasuredMemStats,
+	}
+
+	mu := h.container.CgroupItem("memory.max_usage_in_bytes")
+	for _, rawMemMaxUsage := range mu {
+		val, err := strconv.ParseUint(rawMemMaxUsage, 10, 64)
+		if err != nil {
+			h.logger.Printf("[ERR] driver.lxc: unable to get max memory usage: %v", err)
+			continue
+		}
+		ms.MaxUsage = val
+	}
+	ku := h.container.CgroupItem("memory.kmem.usage_in_bytes")
+	for _, rawKernelUsage := range ku {
+		val, err := strconv.ParseUint(rawKernelUsage, 10, 64)
+		if err != nil {
+			h.logger.Printf("[ERR] driver.lxc: unable to get kernel memory usage: %v", err)
+			continue
+		}
+		ms.KernelUsage = val
+	}
+
+	mku := h.container.CgroupItem("memory.kmem.max_usage_in_bytes")
+	for _, rawMaxKernelUsage := range mku {
+		val, err := strconv.ParseUint(rawMaxKernelUsage, 10, 64)
+		if err != nil {
+			h.logger.Printf("[ERR] driver.lxc: unable to get max kernel memory usage: %v", err)
+			continue
+		}
+		ms.KernelMaxUsage = val
+	}
+
+	taskResUsage := cstructs.TaskResourceUsage{
+		ResourceUsage: &cstructs.ResourceUsage{
+			CpuStats:    cs,
+			MemoryStats: ms,
+		},
+		Timestamp: t.UTC().UnixNano(),
+	}
+
+	return &taskResUsage, nil
+}
+
+func (h *lxcDriverHandle) run() {
+	defer close(h.waitCh)
+	timer := time.NewTimer(containerMonitorIntv)
+	for {
+		select {
+		case <-timer.C:
+			process, err := os.FindProcess(h.initPid)
+			if err != nil {
+				h.waitCh <- &dstructs.WaitResult{Err: err}
+				return
+			}
+			if err := process.Signal(syscall.Signal(0)); err != nil {
+				h.waitCh <- &dstructs.WaitResult{}
```

Comment:
@schmichael This is a cheap trick to know if a process is alive on Linux. So this is not really an error in the container, we are just checking if the init-pid of the container is still alive.